### PR TITLE
Add semicolon inside/outside block lints

### PR DESCRIPTION
**Pull request by 1c3t3a**
This PR adds two new lints to clippy according to #7322. The two lints offer two flavors of the same thing: they check if for `()` returning blocks the semicolon is inside or outside the block and correct it if possible. At the moment the `semicolon_outside_block` lint fires in the `dogfood` test.

- \[x] Followed [lint naming conventions][lint_naming]
- \[x] Added passing UI tests (including committed `.stderr` file)
- \[ ] `cargo test` passes locally
- \[x] Executed `cargo dev update_lints`
- \[x] Added lint documentation
- \[x] Run `cargo dev fmt`

changelog:
Added two new lints according to #7322.
``[`semicolon_inside_block`]``: Checks for a block with a `()` returning expression if the semicolon is inside the block.
``[`semicolon_outside_block`]``: Checks for a block with a `()` returning expression if the semicolon is outside the block.

### CODE REVIEW

**Comment by 1c3t3a** [comment]
I don't really understand why the `semicolon_outside_block` test is failing. On my machine the test runs through and the suite errors on the dogfood tests...

**Comment by xFrednet** [comment]
I believe that the pipeline rebases your change on the current master that might be the reason for the failing test. Have you tried rebasing your changes? :upside_down_face:

**Comment by 1c3t3a** [comment]
I refactored the `semicolon_outside_block` lint to only use `check_fn` instead of the `parent_iter` and `HirId` comparison hocus pocus. It still fires inside the clippy codebase in situations where I personnally would not apply the lint. Example:
```
error: consider moving the `;` outside the block for consistent formatting
1072
   --> src/driver.rs:99:69
1073
    |
1074
99  |           config.parse_sess_created = Some(Box::new(move |parse_sess| {
1075
    |  _____________________________________________________________________^
1076
100 | |             track_clippy_args(parse_sess, &clippy_args_var);
1077
101 | |         }));
1078
    | |____________^
1079
    |
1080
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#semicolon_outside_block
1081
help: put the `;` outside the block
1082
    |
1083
99  ~         config.parse_sess_created = Some(Box::new(move |parse_sess| {
1084
100 +             track_clippy_args(parse_sess, &clippy_args_var)
1085
101 +         }));
1086
    |
```

I think it might be a good idea to move the lint into the `restriction` category or `#[allow()]`ing it.

**Comment by flip1995** [comment]
> I think it might be a good idea to move the lint into the `restriction` category or `#[allow()]`ing it.

If it's 2 lints that contradict themselves, then they _have_ to go into `restriction`, since projects (like Clippy), that enable the whole `pedantic` group will get contradicting or over-eager suggestions.

Another possibility would be to merge those 2 lints into one lint with a configuration option.  But in any case, I see this as a `restriction` lint.

I'm not sure what the better approach is here. I'm fine with continuing with 2 separate lints for now and make the final decision right before the PR is merged.

---

Another thing I noticed when skimming this PR is that those 2 lints are in separate files / lint passes. I would expect that almost all of the code for those 2 lints could be shared, except for the final decision what lint should be triggered. So what's the motivation in splitting this up?

File: clippy_lints/src/semicolon_outside_block.rs [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    /// ```
+    pub SEMICOLON_OUTSIDE_BLOCK,
+    pedantic,
+    "add a semicolon outside the block"
+}
+
+declare_lint_pass!(SemicolonOutsideBlock => [SEMICOLON_OUTSIDE_BLOCK]);
+
+impl LateLintPass<'_> for SemicolonOutsideBlock {
+    fn check_fn(
```
Comment (flip1995) [inline clippy_lints/src/semicolon_outside_block.rs:41]: Why do we have to check a whole function here? Couldn't you just use `check_block` and then get the parent stmt and check if it is a `StmtKind::Semi`?

**Comment by giraffate** [comment]
ping from triage @1c3t3a. Can you have any updates on this? If you have any questions on this, feel free to ask us.

**Comment by 1c3t3a** [comment]
@giraffate Hei there, I am still planning to implement this, but I probably won't have time to do so until October. Is that fine with you? Otherwise I am ok with someone taking over.

**Comment by giraffate** [comment]
@1c3t3a 
> Hei there, I am still planning to implement this, but I probably won't have time to do so until October. Is that fine with you?

Okay!

**Comment by giraffate** [comment]
ping from triage @1c3t3a. Can you have time to work on this?

**Comment by flip1995** [comment]
Also if you have any questions blocking you from continue working on this, don't hesitate to ask!

**Comment by giraffate** [comment]
@1c3t3a ping from triage. According the triage procedure, I'm closing this because 2 weeks have passed with no activity. If you have more time to work on this, feel free to reopen this.

**Comment by Veykril** [comment]
@1c3t3a Would you mind if I try to pick this PR up again bringing it over the finish line?

**Comment by 1c3t3a** [comment]
Sure go ahead! I don't plan to work on this in the next time :)

**Pull request closed without merging**